### PR TITLE
Zip code textfield

### PR DESCRIPTION
**Pull request by joesus**

### CODE REVIEW

File: AllisonsCatPalaceTDDTests/AnimalDetailHeaderViewTests.swift [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        let header = AnimalDetailHeaderView(frame: frame)
+
+        setNeedsLayoutSpy = UIView.SetNeedsLayoutSpyController.createSpy(on: header)
+        setNeedsLayoutSpy?.beginSpying()
+
+        let setNeedsLayoutCalledExpectation = NSPredicate { _, _ in
+            header.setNeedsLayoutCalled
+        }
+
+        let _ = expectation(for: setNeedsLayoutCalledExpectation, evaluatedWith: self, handler: nil)
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/AnimalDetailHeaderViewTests.swift:86]: don't need let _ =

Comment (joesus) [reply AllisonsCatPalaceTDDTests/AnimalDetailHeaderViewTests.swift:86]: also don't need handler: nil

File: AllisonsCatPalaceTDDTests/CatNetworkerTests.swift [hunk 2]
```diff
@@ -51,14 +51,45 @@ class PetFinderNetworkerTests: XCTestCase {
         XCTAssertEqual(request.url?.path, "/pet.find", "The path should be cats")
         XCTAssertTrue(request.url!.query!.contains("format=json"), "Query: \(request.url!.query!) should specify json format response")
         XCTAssertTrue(request.url!.query!.contains("output=full"), "Query: \(request.url!.query!) should specify output size")
-        XCTAssertTrue(request.url!.query!.contains("location=80220"), "Query: \(request.url!.query!) should specify location")
+        XCTAssertTrue(request.url!.query!.contains("location="), "Query: \(request.url!.query!) should specify location")
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/CatNetworkerTests.swift:54]: call to retrieveAllAnimals should be passed dependencies (zip code)

File: AllisonsCatPalaceTDDTests/LocationControllerTests.swift [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    }
+
+    func testViewDidAppear() {
+        UIViewController.ViewDidAppearSpyController.createSpy(on: controller)!.spy {
+            replaceRootViewController(with: controller)
+            controller.viewDidAppear(false)
+            XCTAssert(controller.superclassViewDidAppearCalled,
+                      "ViewDidAppear should call viewDidAppear on controller")
+
+            XCTAssertTrue(textField.isFirstResponder,
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/LocationControllerTests.swift]: This should be it's own test

File: AllisonsCatPalaceTDDTests/LocationControllerTests.swift [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        XCTAssertEqual(textField.layer.borderColor, UIColor.green.cgColor,
+                       "Zip code field's border color should be green")
+
+        delegate.textFieldDidBeginEditing?(textField)
+        XCTAssertEqual(textField.layer.borderColor, UIColor.lightGray.cgColor,
+                       "Zip code field's border color should be reset to light gray on beginning text editing")
+
+        // stops spinner
+        controller.activityIndicator.startAnimating()
+        XCTAssertTrue(controller.activityIndicator.isAnimating,
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/LocationControllerTests.swift]: take this out. Will know that it's animating if I told UIKit to start animating.

File: AllisonsCatPalaceTDDTests/LocationControllerTests.swift [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+            // Need to be in the window to becomeFirstResponder
+            replaceRootViewController(with: controller)
+            textField.becomeFirstResponder()
+            XCTAssertTrue(textField.isEditing,
+                          "Textfield should be editing when it's the first responder")
+
+            textField.text = "80220"
+            controller.textFieldDidChange(textField)
+
+            XCTAssertFalse(textField.isEditing,
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/LocationControllerTests.swift]: Don't need to test this because it's implicit if first responder is resigned.

File: AllisonsCatPalaceTDDTests/LocationControllerTests.swift [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+                           "Textfield should not be editing when it is not the first responder")
+            XCTAssertTrue(textField.resignFirstResponderCalled,
+                          "Textfield should resign first responder at five characters")
+        }
+    }
+
+    func testZipCodeFieldResignsFirstResponderWhenDoneEditing() {
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/LocationControllerTests.swift]: no reason for this since didEndEditing is called when first responder is resigned.

File: AllisonsCatPalaceTDDTests/LocationControllerTests.swift [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+            XCTAssertTrue(textField.resignFirstResponderCalled,
+                          "Textfield should resign first responder when done editing")
+        }
+    }
+
+    func testZipCodeFieldGeocodesAtFiveCharacters() {
+        textField.text = "80220"
+        delegate.textFieldDidEndEditing!(textField)
+        XCTAssertTrue(geocoder.forwardGeocodeAddressCalled)
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/LocationControllerTests.swift]: Needs message.

File: AllisonsCatPalaceTDDTests/LocationControllerTests.swift [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+    }
+
+    func testActivityIndicatorHiddenByDefault() {
+        XCTAssertTrue(controller.activityIndicator.isHidden,
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/LocationControllerTests.swift]: just test if it's hidden and if it hides when stopped.

File: AllisonsCatPalaceTDDTests/LocationControllerTests.swift [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+    func testGeocodingCompleted() {
+        textField.text = "80220"
+        delegate.textFieldDidEndEditing!(textField)
+
+        guard let handler = geocoder.forwardGeocodeAddressCompletionHandler else {
+            return XCTFail("Geocoder should be called with a handler")
+        }
+        handler([], nil)
+
+        XCTAssertTrue(controller.activityIndicator.isHidden,
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/LocationControllerTests.swift]: just make sure that it's stopped since an earlier test checks hides when stopped.

File: AllisonsCatPalaceTDDTests/LocationControllerTests.swift [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+                      "Activity indicator should be hidden when geocoding is not in progress")
+        XCTAssertFalse(controller.activityIndicator.isAnimating,
+                       "Activity indicator should not be animating when geocoding is not in progress")
+        XCTAssertTrue(textField.isUserInteractionEnabled,
+                      "User interaction should be enabled when geocoding is not in progress")
+    }
+
+
+    func testGeocodingWithEmptyResultsAndError() {
+        delegate.textFieldDidEndEditing!(textField)
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/LocationControllerTests.swift]: fix by triggering directly.

File: AllisonsCatPalaceTDDTests/LocationControllerTests.swift [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+            return XCTFail("Geocoder should be called with a handler")
+        }
+        handler([], nil)
+
+        XCTAssertEqual(textField.layer.borderColor, UIColor.red.cgColor,
+                       "Zip code field should have red border when there are no results")
+        // TODO: use custom error for empty results to show the message
+    }
+
+    func testGeocodingWithNonEmptyResultsAndError() {
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/LocationControllerTests.swift]: This won't be possible because of CoreLocation. Safe assumption to make.

File: AllisonsCatPalaceTDDTests/LocationControllerTests.swift [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+        showSpy?.beginSpying()
+
+        delegate.textFieldDidEndEditing!(textField)
+
+        guard let handler = geocoder.forwardGeocodeAddressCompletionHandler else {
+            return XCTFail("Geocoder should be called with a handler")
+        }
+        handler([placemark], nil)
+
+        XCTAssertTrue(navController.showCalled,
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/LocationControllerTests.swift:260]: Test the segue by finding it in the storyboard and then test that it's a show and that it has the correct destination.

File: AllisonsCatPalaceTDDTests/LocationControllerTests.swift [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+
+        guard let handler = geocoder.forwardGeocodeAddressCompletionHandler else {
+            return XCTFail("Geocoder should be called with a handler")
+        }
+        handler([placemark], nil)
+
+        XCTAssertEqual(SettingsManager.shared.value(forKey: SettingsManager.Key.zipCode) as? String, "80220",
+                       "Geocoding should save zip code to settings")
+    }
+
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/LocationControllerTests.swift:287]: Add test that textField is empty when persisted zip code is empty

File: AllisonsCatPalaceTDDTests/CatListControllerTests.swift [hunk 14]
```diff
@@ -140,4 +122,30 @@ class CatListControllerTests: XCTestCase {
...
+        }
+        expectation(for: NSPredicate(block: predicateBlock), evaluatedWith: self)
+
+        UIViewController.PerformSegueSpyController.createSpy(on: controller)!.spy {
+            controller.cats = [SampleCat]
+
+            let cell = controller.tableView.cellForRow(at: firstCatIndexPath) as? CatCell
+            controller.performSegue(withIdentifier: "ShowCatDetail", sender: cell)
+
+            waitForExpectations(timeout: 200, handler: nil)
```
Comment (joesus) [inline AllisonsCatPalaceTDDTests/CatListControllerTests.swift]: fix timeout

File: AllisonsCatPalaceTDD/LocationController.swift [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    @IBOutlet weak var activityIndicator: UIActivityIndicatorView!
+
+    var geocoder = CLGeocoder()
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        decorateZipCodeField()
+        populateZipCodeField()
+        zipCodeField.addTarget(self, action: #selector(textFieldDidChange(_:)), for: .editingChanged)
```
Comment (joesus) [inline AllisonsCatPalaceTDD/LocationController.swift]: fully qualified selectors; ie: LocationController.textFieldDidChange(_:)
also use .valueChanged vs .editingChanged

File: AllisonsCatPalaceTDD/LocationController.swift [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+
+    fileprivate func decorateZipCodeField() {
+        zipCodeField.layer.cornerRadius = 4
+        zipCodeField.layer.borderWidth = 1
+        zipCodeField.layer.borderColor = UIColor.lightGray.cgColor
+    }
+
+    fileprivate func populateZipCodeField() {
+        let storedZipCode = SettingsManager.shared.value(forKey: .zipCode) as? String
+        zipCodeField.text = storedZipCode ?? ""
```
Comment (joesus) [inline AllisonsCatPalaceTDD/LocationController.swift]: don't need ?? ""

File: AllisonsCatPalaceTDD/LocationController.swift [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        }
+    }
+}
+
+extension LocationController: UITextFieldDelegate {
+
+    func textFieldDidBeginEditing(_ textField: UITextField) {
+        textField.layer.borderColor = UIColor.lightGray.cgColor
+
+        activityIndicator.stopAnimating() // find out if it's bad to stop if already stopped
```
Comment (joesus) [inline AllisonsCatPalaceTDD/LocationController.swift]: remove comment

File: AllisonsCatPalaceTDD/LocationController.swift [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+        let originalString = textField.text ?? ""
+        let newString = (originalString as NSString).replacingCharacters(in: range, with: string)
+        let newStringOnlyContainsDecimals = CharacterSet.decimalDigits.isSuperset(of: CharacterSet(charactersIn: newString))
+
+        guard newStringOnlyContainsDecimals else {
+            return false
+        }
+
+        if newString.characters.count > 5 {
+            textField.endEditing(true)
```
Comment (joesus) [inline AllisonsCatPalaceTDD/LocationController.swift]: remove keyboard hiding.

File: AllisonsCatPalaceTDD/LocationController.swift [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+        let originalString = textField.text ?? ""
+        let newString = (originalString as NSString).replacingCharacters(in: range, with: string)
+        let newStringOnlyContainsDecimals = CharacterSet.decimalDigits.isSuperset(of: CharacterSet(charactersIn: newString))
+
+        guard newStringOnlyContainsDecimals else {
+            return false
+        }
+
+        if newString.characters.count > 5 {
```
Comment (joesus) [inline AllisonsCatPalaceTDD/LocationController.swift]: change to be exactly five and resignFirstResponder to trigger geocoding. 
Also if it's >5 return false.

File: AllisonsCatPalaceTDD/LocationController.swift [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+            textField.endEditing(true)
+            return false
+        }
+
+        return true
+    }
+
+    func textFieldDidEndEditing(_ textField: UITextField) {
+        textField.resignFirstResponder()
```
Comment (joesus) [inline AllisonsCatPalaceTDD/LocationController.swift]: resignFirstResponder() will call this. Don't call this or endEditing() directly. Then there's no need to resign it here since that's the catalyst for this.

File: AllisonsCatPalaceTDD/LocationController.swift [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+
+        activityIndicator.startAnimating()
+        textField.isUserInteractionEnabled = false
+
+        geocoder.geocodeAddressString(zipCodeText) { [weak self] (placemarks, error) in
```
Comment (joesus) [inline AllisonsCatPalaceTDD/LocationController.swift]: move into own method that takes the zipcode.

File: AllisonsCatPalaceTDD/AnimalNetworker.swift [hunk 22]
```diff
@@ -34,15 +34,22 @@ enum PetFinderNetworker {
...
+        let location = SettingsManager.shared.value(forKey: .zipCode) as? String ?? ""
+
         retrieveAllAnimalsTask?.cancel()
 
         CatServiceComponents.path = "/pet.find"
         let locationQuery = URLQueryItem(name: "location", value: location)
-        if CatServiceComponents.queryItems?.contains(locationQuery) == false {
-            CatServiceComponents.queryItems?.append(locationQuery)
+
+        if let oldLocationQuery = CatServiceComponents.queryItems?.first(where: { $0.name == "location" }),
```
Comment (joesus) [inline AllisonsCatPalaceTDD/AnimalNetworker.swift]: this can go once urlComponents are being created in-line.

File: AllisonsCatPalaceTDD/AnimalDetailHeaderView.swift [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+@IBDesignable
+class AnimalDetailHeaderView: UIView {
+    @IBOutlet weak var imageView: UIImageView!
+
+    var animal: Animal? {
+        didSet {
+            guard let url = animal?.imageLocations.large.first else {
+                return
+            }
+            ImageProvider.getImage(for: url) { [weak self] image in
```
Comment (joesus) [inline AllisonsCatPalaceTDD/AnimalDetailHeaderView.swift]: doesn't need to be weak self
